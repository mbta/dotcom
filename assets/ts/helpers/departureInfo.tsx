import React from "react";
import {
  chain,
  concat,
  difference,
  find,
  keyBy,
  keys,
  map,
  sortBy
} from "lodash";
import { PredictionWithTimestamp } from "../models/predictions";
import { ScheduleWithTimestamp } from "../models/schedules";
import { isCancelled, isDelayed, isSkipped } from "./prediction-helpers";
import { DepartureInfo } from "../models/departureInfo";
import {
  isABusRoute,
  isACommuterRailRoute,
  isSubwayRoute
} from "../models/route";
import { DirectionId, Route } from "../__v3api";
import DisplayTime from "../stop/components/DisplayTime";
import { getInfoKey } from "../stop/models/displayTimeConfig";
import { RoutePatternWithPolyline } from "../models/route-patterns";

export const SUBWAY = "subway";
export const BUS = "bus";
export const COMMUTER_RAIL = "commuter_rail";
export const FERRY = "ferry";

const toRouteMode = (
  route: Route
): typeof SUBWAY | typeof BUS | typeof COMMUTER_RAIL | typeof FERRY => {
  if (isSubwayRoute(route)) {
    return SUBWAY;
  }
  if (isABusRoute(route)) {
    return BUS;
  }
  if (isACommuterRailRoute(route)) {
    return COMMUTER_RAIL;
  }
  return FERRY;
};

const departureInfoToTime = (
  departureInfo: DepartureInfo
): Date | undefined => {
  return departureInfo.prediction?.time || departureInfo.schedule?.time;
};

const getNextUnCancelledDeparture = (
  departureInfos: DepartureInfo[]
): DepartureInfo | undefined => {
  return find(departureInfos, (di: DepartureInfo) => !di.isCancelled);
};

const displayInfoContainsPrediction = (
  departureInfo: DepartureInfo
): boolean => {
  return !!departureInfo.prediction;
};

const mergeIntoDepartureInfo = (
  schedules: ScheduleWithTimestamp[],
  predictions: PredictionWithTimestamp[]
): DepartureInfo[] => {
  const schedulesByTripId = keyBy(schedules, "trip.id");
  const predictionsByTripId = keyBy(predictions, "trip.id");
  const scheduleTripIds = keys(schedulesByTripId);
  const predictionTripIds = keys(predictionsByTripId);

  // All the schedules with and without predictions
  const scheduleDirectionInfo = map(
    schedulesByTripId,
    (schedule, tripId): DepartureInfo => {
      // prediction could be undefined
      const prediction = predictionsByTripId[tripId];
      return {
        prediction,
        schedule,
        route: schedule.route,
        trip: schedule.trip,
        isCancelled: isCancelled(prediction),
        isSkipped: isSkipped(prediction),
        isDelayed: isDelayed(prediction, schedule),
        routeMode: toRouteMode(schedule.route)
      };
    }
  );

  const unmatchedPredictionTripIds = difference(
    predictionTripIds,
    scheduleTripIds
  );
  // All the predictions with no schedules
  const predictionDirectionInfo = map(
    unmatchedPredictionTripIds,
    (tripId: string): DepartureInfo => {
      const prediction = predictionsByTripId[tripId];
      return {
        prediction,
        route: prediction.route,
        trip: prediction.trip,
        isCancelled: isCancelled(prediction),
        isSkipped: isSkipped(prediction),
        routeMode: toRouteMode(prediction.route)
      } as DepartureInfo;
    }
  );

  const departureInfos = concat(scheduleDirectionInfo, predictionDirectionInfo);
  const sortedDepartureInfo = sortBy(departureInfos, di => {
    // prioritize sorting by predictions
    if (di.prediction && di.prediction.time) {
      return di.prediction.time;
    }
    return di.schedule?.time;
  });

  return sortedDepartureInfo;
};

const isAtDestination = (
  stopName: string,
  route: Route,
  directionId: DirectionId
): boolean => {
  const destination = route.direction_destinations[directionId];
  if (destination) {
    return destination.includes(stopName) && !destination.includes(" or ");
  }
  return false;
};

const isCommuterRail = (departureInfo: DepartureInfo): boolean => {
  return departureInfo?.routeMode === "commuter_rail";
};

const DefaultWrapper: React.FunctionComponent<unknown> = ({ children }) => (
  <li>{children}</li>
);
/**
 * From a list of `DepartureInfo[]`, generate a list of `<DisplayTime />`.
 * Each `<DisplayTime />` will recieve the `isCR` and `targetDate` props.
 *
 * This function can optionally
 * - wrap each `<DisplayTime />` in a specified wrapper element, `WrapperEl`
 * - truncate the list to a desired `listLength` value
 */
const departuresListFromInfos = (
  departureInfos: DepartureInfo[],
  isCR: boolean,
  isSubway: boolean,
  targetDate?: Date,
  listLength?: number,
  omitCancelledAndSkipped = false,
  WrapperEl: typeof DefaultWrapper = DefaultWrapper
): React.ReactElement[] => {
  const predictions = chain(departureInfos)
    .reject(
      departure =>
        (omitCancelledAndSkipped &&
          (!!departure.isCancelled || !!departure.isSkipped)) ||
        (isSubway && typeof departure.prediction === "undefined")
    )
    .slice(0, listLength)
    .value();

  if (predictions.length === 0) {
    const routeId = departureInfos[0]?.route?.id;
    const tripId = departureInfos[0]?.trip?.id;

    return [
      <div className="text-sm font-helvetica" key={`${routeId}-${tripId}`}>
        No real-time data
      </div>
    ];
  }

  return predictions.map(d => (
    <WrapperEl key={getInfoKey(d)}>
      <DisplayTime departure={d} isCR={isCR} targetDate={targetDate} />
    </WrapperEl>
  ));
};

const departureInfoInRoutePatterns = (
  departureInfo: DepartureInfo,
  routePatterns: RoutePatternWithPolyline[]
): boolean => {
  const isMatch = !!routePatterns.find(
    rp => rp.id === departureInfo.trip.route_pattern_id
  );
  // This second case is needed for the edge case where the route pattern is
  // canonical but the schedules/predictions are not serving canonical route
  // patterns (e.g. due to shuttles or other disruptions)
  const isSimilar = !!routePatterns.find(
    rp =>
      rp.headsign === departureInfo.trip.headsign &&
      rp.route_id === departureInfo.route.id
  );
  return isMatch || isSimilar;
};

export {
  mergeIntoDepartureInfo,
  departureInfoToTime,
  displayInfoContainsPrediction,
  getNextUnCancelledDeparture,
  isAtDestination,
  isCommuterRail,
  departuresListFromInfos,
  departureInfoInRoutePatterns
};
