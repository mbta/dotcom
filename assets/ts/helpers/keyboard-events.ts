export const isEnter = (key: number | string): boolean =>
  key === "Enter" || key === 13;

export const isEscape = (key: number | string): boolean =>
  key === "Escape" || key === 27;

// for vanilla JS
export const handleNativeEnterKeyPress = (
  e: KeyboardEvent,
  cb: Function
): void => (isEnter(e.key || e.keyCode) ? cb(e) : (() => {})());

export const handleNativeEscapeKeyPress = (
  e: KeyboardEvent,
  cb: Function
): void => (isEscape(e.key || e.keyCode) ? cb(e) : (() => {})());
