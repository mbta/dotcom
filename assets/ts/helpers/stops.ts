import { Stop } from "../__v3api";
// eslint-disable-next-line import/prefer-default-export
export const isStopAStation = (stop: Stop): boolean => stop["station?"];

export const isStopAFerryStop = (stop: Stop): boolean => stop["ferry?"];
