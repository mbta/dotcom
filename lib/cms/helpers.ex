defmodule CMS.Helpers do
  @moduledoc """
  Various helper functions that aid in parsing CMS JSON data.
  """

  alias CMS.API
  alias CMS.CustomHTML5Scrubber
  alias CMS.Field.{File, Image, Link}
  alias CMS.Partial.Paragraph
  alias Phoenix.HTML

  @static_path "/sites/default/files"

  @doc """
  Each CMS entity hosts a variety of fields which all
  follow common patterns, depending on the field type:
  - Long text field (optional: with summary):
    - value: raw database value from user input
    - processed: Drupal's post-processed (safer) value
    - summary: A plain-text value separate from above
  - Entity reference field (collapsed)
    - target_id: a integer value of another entity
  - Entity reference field (expanded)
    - the entire map is the page struct
  - default: most other fields
    - value: mixed type, but most common pattern
  """
  @spec field_value(map, String.t()) :: any
  def field_value(entity, field) do
    case entity[field] do
      [%{"processed" => value}] -> value
      [%{"value" => value}] -> value
      [%{"target_id" => value}] -> value
      [%{"nid" => _} = node] -> node
      _ -> nil
    end
  end

  @doc """
  Handles entity fields that support multiple values.
  Common patterns above apply here.
  """
  @spec field_values(map, String.t()) :: [map]
  def field_values(entity, field) do
    entity
    |> Map.get(field, [])
    |> Enum.map(&Map.get(&1, "value"))
  end

  @spec handle_html(String.t() | nil) :: HTML.safe()
  def handle_html(html) do
    (html || "")
    |> CustomHTML5Scrubber.html5()
    |> rewrite_static_file_links
    |> HTML.raw()
  end

  @spec parse_body(map) :: HTML.safe()
  def parse_body(%{} = data) do
    data
    |> field_value("body")
    |> handle_html
  end

  @spec parse_files(map, String.t()) :: [File.t()]
  def parse_files(%{} = data, field) do
    data
    |> Map.get(field, [])
    |> Enum.map(&File.from_api/1)
  end

  @spec parse_page_types(map) :: list(String.t())
  def parse_page_types(%{} = data) do
    data
    |> Map.get("field_page_type", [])
    |> Kernel.get_in([Access.all(), "name"])
    |> Enum.reject(&Kernel.is_nil/1)
  end

  @spec parse_related_transit(map) :: list(String.t())
  def parse_related_transit(%{} = data) do
    data
    |> Map.get("field_related_transit", [])
    |> Kernel.get_in([Access.all(), "name"])
    |> Enum.reject(&Kernel.is_nil/1)
  end

  @spec path_alias(map) :: String.t() | nil
  def path_alias(data) do
    data
    |> parse_path_alias()
  end

  @spec parse_path_alias(map) :: String.t() | nil
  def parse_path_alias(%{"path" => [%{"alias" => path_alias}]}), do: path_alias
  def parse_path_alias(_), do: nil

  @spec parse_image(map, String.t()) :: Image.t() | nil
  def parse_image(%{} = data, field) do
    case parse_images(data, field) do
      [image] -> image
      [] -> nil
    end
  end

  @spec parse_images(map, String.t()) :: [Image.t()] | []
  def parse_images(%{} = data, field) do
    data
    |> Map.get(field, [])
    |> Enum.map(&Image.from_api/1)
  end

  @spec parse_iso_datetime(String.t()) :: DateTime.t() | nil
  def parse_iso_datetime(nil) do
    nil
  end

  def parse_iso_datetime(time) do
    case String.split(time, ":") do
      [_date_hr, _min, _sec] ->
        time
        |> Timex.parse("{ISOdate}T{ISOtime}")
        |> do_parse_iso_datetime(:deprecated)

      [_date_hr, _min, _sec, _tz] ->
        time
        |> Timex.parse("{ISO:Extended}")
        |> do_parse_iso_datetime(:extended)

      _ ->
        nil
    end
  end

  defp do_parse_iso_datetime({:ok, dt}, :deprecated), do: Timex.to_datetime(dt, "Etc/UTC")
  defp do_parse_iso_datetime({:ok, dt}, :extended), do: dt
  defp do_parse_iso_datetime(_, _), do: nil

  @spec parse_date(map, String.t()) :: Date.t() | nil
  def parse_date(data, field) do
    case data[field] do
      [%{"value" => date}] -> parse_date_string(date, "{YYYY}-{0M}-{0D}")
      _ -> nil
    end
  end

  @spec parse_date_string(String.t(), String.t()) :: Date.t() | nil
  defp parse_date_string(date, format_string) do
    case Timex.parse(date, format_string) do
      {:error, _message} -> nil
      {:ok, naive_datetime} -> NaiveDateTime.to_date(naive_datetime)
    end
  end

  @spec parse_link(map, String.t()) :: Link.t() | nil
  def parse_link(%{} = data, field) do
    case data[field] do
      [link] -> Link.from_api(link)
      _ -> nil
    end
  end

  # This is silly and should be combined/refactored with above clause
  @spec parse_links(map, String.t()) :: [Link.t()] | nil
  def parse_links(%{} = data, field) do
    case data[field] do
      [_ | _] = links -> Enum.map(links, &Link.from_api(&1))
      _ -> nil
    end
  end

  @doc """
  Gathers information about desired view mode based on incoming
  query parameters. Key presence is enough to count as true.
  - page: whether the page is in preview mode (drafts are rendered)
  - paragraphs: whether unpublished paragraphs should render
  """
  @spec preview_opts(map) :: Keyword.t()
  def preview_opts(query_params \\ %{}) do
    [
      page: Map.has_key?(query_params, "preview"),
      paragraphs: Map.has_key?(query_params, "paragraphs")
    ]
  end

  @doc """
  Expects raw JSON data for a CMS object that contains a paragraphs field.
  This field value will always be a list of potential paragraphs.
  """
  @spec parse_paragraphs(map, Keyword.t(), String.t()) :: [Paragraph.t()]
  def parse_paragraphs(data, preview_opts \\ [], target_field \\ "field_paragraphs") do
    data
    |> Map.get(target_field, [])
    |> Enum.filter(&show_paragraph?(&1, preview_opts))
    |> Enum.map(&Paragraph.from_api(&1, preview_opts))
  end

  @spec show_paragraph?(map | nil, Keyword.t()) :: boolean
  defp show_paragraph?(field_data, preview_opts)

  # Skip broken/missing paragraphs (CMS unable to load and returns NULL)
  defp show_paragraph?(nil, _) do
    false
  end

  # Reusable paragraphs instance aren't automatically removed when their child
  # paragraphs are deleted from the database, so catch that here.
  defp show_paragraph?(%{"field_reusable_paragraph" => [nil]}, _) do
    false
  end

  # Reusable paragraphs are not directly renderable since they act as instance containers.
  # However, these instances can be unpublished. If unpublished, stop and return false.
  # If published, continue checking the nested child paragraph for publish status.
  defp show_paragraph?(%{"field_reusable_paragraph" => [child]} = parent, preview_opts) do
    %{"status" => [parent_status]} = parent
    %{"paragraphs" => [paragraph]} = child

    case parent_status do
      %{"value" => false} -> false
      _ -> show_paragraph?(paragraph, preview_opts)
    end
  end

  # In "preview" mode, allow unpublished paragraphs to be rendered if requested
  defp show_paragraph?(_, preview: true, paragraphs: true) do
    true
  end

  defp show_paragraph?(%{"status" => [%{"value" => value}]}, _) do
    value
  end

  @spec rewrite_static_file_links(String.t()) :: String.t()
  defp rewrite_static_file_links(body) do
    if Application.get_env(:dotcom, :is_prod_env?) do
      Regex.replace(~r/"(#{@static_path}[^"]+)"/, body, fn _, path ->
        [~c"\"", Util.site_path(:static_url, [path]), ~c"\""]
      end)
    else
      body
    end
  end

  @spec rewrite_url(String.t()) :: String.t()
  def rewrite_url(url) when is_binary(url) do
    uri = URI.parse(url)

    path =
      if uri.query do
        "#{uri.path}?#{uri.query}"
      else
        uri.path
      end

    if Application.get_env(:dotcom, :is_prod_env?) do
      Util.site_path(:static_url, [path])
    else
      path
    end
  end

  @spec int_or_string_to_int(integer | String.t() | nil) :: integer | nil
  def int_or_string_to_int(nil), do: nil
  def int_or_string_to_int(num) when is_integer(num), do: num

  def int_or_string_to_int(str) when is_binary(str) do
    case Integer.parse(str) do
      {int, ""} -> int
      _ -> nil
    end
  end

  @doc """
  Retrieves category from CMS data. If "field_page_type" field is
  empty or not found, the returned category is an empty string.

  iex> category(%{"field_page_type" => [%{"name" => "Guides"}]})
  "Guides"
  iex> category(%{"field_page_type" => []})
  ""
  iex> category(%{})
  ""
  iex> category(nil)
  ""
  """
  @spec category(map | nil) :: String.t()
  def category(nil), do: ""

  def category(data) do
    data
    |> Map.get("field_page_type", [%{}])
    |> Enum.at(0, %{})
    |> Map.get("name", "")
  end

  @spec content_type(String.t()) :: API.type() | nil
  for atom <- ~w(
    diversion
    event
    news_entry
    page
    project
    project_update
    )a do
    str = Atom.to_string(atom)
    def content_type(unquote(str)), do: unquote(atom)
  end

  def content_type(_), do: nil

  @doc "Returns the text if present, otherwise returns nil"
  @spec content(String.t()) :: String.t() | nil
  @spec content(HTML.safe()) :: HTML.safe() | nil
  def content(nil) do
    nil
  end

  def content({:safe, string} = safe_html) do
    if content(string) do
      safe_html
    end
  end

  def content(string) do
    case String.trim(string) do
      "" -> nil
      string -> string
    end
  end

  @doc """
  Parses the related_transit field for route-specific data. Could
  contain multiple routes. Contains mode, branch, and other info.
  """
  @spec routes([map()]) :: [API.route_term()]
  def routes(route_data) do
    route_data
    |> Enum.map(&Map.get(&1, "data"))
    |> Enum.map(&route_metadata/1)
  end

  # Maps the tagged CMS route term, its group, and its parent mode.
  # For routes and misc. groupings like "local_bus," the CMS will have
  # mapped the appropriate GTFS mode to that tag prior to parsing here.
  @spec route_metadata(map()) :: API.route_term()
  defp route_metadata(route_data) do
    Map.new(
      id: Map.get(route_data, "gtfs_id"),
      group: Map.get(route_data, "gtfs_group"),
      mode:
        route_data
        |> Map.get("gtfs_ancestry")
        |> Map.get("mode")
        |> route_mode()
    )
  end

  # Some CMS routes are actually custom groups that may not
  # have any single MBTA mode associated with them (mode: nil).
  # There should never be more than one, single mode in the list.
  @spec route_mode([String.t()] | nil) :: String.t() | nil
  defp route_mode(nil), do: nil
  defp route_mode([mode]), do: mode
end
