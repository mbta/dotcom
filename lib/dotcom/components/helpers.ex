defmodule Dotcom.Components.Helpers do
  @doc """
    Takes a string, capitalizes each word between underscores, and removes the underscores
  """
  def alias_name(atom) do
    atom
    |> Kernel.to_string()
    |> String.split("_")
    |> Enum.map_join(&String.capitalize(&1))
  end

  @doc """
    A helper function to ensure that the correct path is returned no
    matter where this module is called from (i.e. using mix test or iex)
  """
  def components_folder_path do
    File.cwd!()
    |> String.split("/lib/dotcom")
    |> List.first()
    |> Path.join("/lib/dotcom/components/")
  end

  def components_section_path(section) when is_atom(section) do
    components_folder_path()
    |> Path.join("#{section}")
  end

  def component_folder_path(component, section) when is_atom(component) and is_atom(section) do
    section
    |> components_section_path()
    |> Path.join("/")
    |> Path.join("#{component}")
  end

  @doc """
    Returns the component's module (i.e. Dotcom.Components.Buttons.ModeButtonList)
  """
  def component_module(component, section) when is_atom(component) and is_atom(section) do
    Dotcom.Components
    |> Module.concat(alias_name(section))
    |> Module.concat(alias_name(component))
  end
end
