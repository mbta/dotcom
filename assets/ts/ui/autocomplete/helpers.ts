import { AutocompleteRenderer } from "@algolia/autocomplete-js";
import { omit } from "lodash";
import { createElement, Fragment } from "react";
import { render } from "react-dom";
import { parseQuery } from "../../helpers/query";
import {
  AutocompleteItem,
  ContentItem,
  Item,
  LocationItem,
  PopularItem,
  RouteItem,
  SearchResultItem,
  StopItem
} from "./__autocomplete";

export function isAlgoliaItem(x: Item): x is AutocompleteItem {
  return Object.keys(x).includes("objectID");
}

export function isStopItem(x: Item): x is StopItem {
  return Object.keys(x).includes("stop");
}

export function isRouteItem(x: Item): x is RouteItem {
  return Object.keys(x).includes("route");
}

export function isContentItem(x: Item): x is ContentItem {
  return Object.keys(x).includes("_content_type");
}

export function isSearchResultItem(x: Item): x is SearchResultItem {
  // eslint-disable-next-line no-underscore-dangle
  return isContentItem(x) && x._content_type === "search_result";
}

export const getTitleAttribute = (item: Item): string[] => {
  if (isStopItem(item)) {
    return ["stop", "name"];
  }
  if (isRouteItem(item)) {
    return ["route", "name"];
  }

  return ["content_title"];
};

export type UrlType =
  | "retail-sales-locations"
  | "proposed-sales-locations"
  | "vote";

export type WithUrls<T> = T & { urls: Record<UrlType, string> };

// The backend returns all possible URLs, use urlType to get the desired one
export const itemWithUrl = (
  initialItem: WithUrls<LocationItem | PopularItem>,
  urlType?: UrlType
): Omit<typeof initialItem, "urls"> => {
  const item = omit(initialItem, "urls");
  if (urlType) {
    item.url = initialItem.urls[urlType];
  }
  return item;
};

export function getLikelyQueryParams(): string | undefined {
  const searchParams = parseQuery(
    window.location.search,
    window.decodeURIComponent
  );
  const { query, name, address, latitude, longitude } = searchParams;
  const latlon =
    latitude && longitude ? `${latitude}, ${longitude}` : undefined;
  return query || name || address || latlon;
}

export const customRenderer = {
  createElement,
  Fragment,
  render
} as AutocompleteRenderer;
