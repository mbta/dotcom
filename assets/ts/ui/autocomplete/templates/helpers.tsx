/* eslint-disable import/prefer-default-export */
/* eslint-disable no-underscore-dangle */
import { SourceTemplates } from "@algolia/autocomplete-js";
import React from "react";
import { AutocompleteItem } from "../__autocomplete";
import { isAlgoliaItem, isContentItem, isSearchResultItem } from "../helpers";
import { LinkForItem } from "./algolia";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ItemTemplate = SourceTemplates<any>["item"];

export const itemURL = (item: AutocompleteItem): string => {
  let url = isContentItem(item) ? item._content_url : item.url;

  // Search result items are a subset of content items that point to a different URL
  if (isSearchResultItem(item)) {
    url = item._search_result_url.replace(/(internal|entity):/g, "/");
  }

  // Strip extra forward slashes as they break relative links
  url = url.replace(/\/\//g, "/");

  return url;
};

export const templateWithLink = (
  childTemplate: ItemTemplate
  // eslint-disable-next-line react/display-name
): ItemTemplate => props => {
  if (isAlgoliaItem(props.item)) {
    return (
      <LinkForItem
        item={props.item as AutocompleteItem}
        query={props.state.query}
      >
        {childTemplate(props)}
      </LinkForItem>
    );
  }
  return (
    <a href={props.item.url} className="aa-ItemLink">
      {childTemplate(props)}
    </a>
  );
};
