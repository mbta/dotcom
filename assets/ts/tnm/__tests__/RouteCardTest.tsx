import React from "react";
import renderer from "react-test-renderer";
import RouteCard from "../components/RouteCard";
import { createReactRoot } from "../../app/helpers/testUtils";
import {
  Direction,
  Headsign,
  PredictedOrScheduledTime,
  Stop,
  RouteWithStopsWithDirections,
  Alert
} from "../../__v3api";

/* eslint-disable camelcase */

const time: PredictedOrScheduledTime = {
  scheduled_time: ["4:30", " ", "PM"],
  prediction: null,
  delay: 0
};

const headsign: Headsign = {
  headsign: "Headsign",
  name: "Headsign - Extra info",
  times: [time],
  train_number: null
};

const directions: Direction[] = [
  {
    direction_id: 0,
    headsigns: [headsign]
  }
];

const stop: Stop = {
  accessibility: ["wheelchair"],
  address: "123 Main St., Boston MA",
  bike_storage: [],
  closed_stop_info: null,
  fare_facilities: [],
  "has_charlie_card_vendor?": false,
  "has_fare_machine?": false,
  id: "stop-id",
  "child?": false,
  latitude: 41.0,
  longitude: -71.0,
  municipality: "Boston",
  name: "Stop Name",
  note: null,
  parking_lots: [],
  "station?": true,
  "ferry?": false,
  href: "/stops/stop-id",
  type: "station"
};

const alert = {} as Alert;

const route: RouteWithStopsWithDirections = {
  route: {
    alerts: [alert],
    direction_destinations: ["Outbound Destination", "Inbound Destination"],
    direction_names: ["Outbound", "Inbound"],
    id: "route-id",
    name: "Route Name",
    header: "Route Header",
    long_name: "Route Long Name",
    description: "Route Description",
    type: 3,
    line_id: null
  },
  stops_with_directions: [{ stop, directions, distance: "238 ft" }]
};

it("it renders a stop card", () => {
  createReactRoot();
  const tree = renderer
    .create(<RouteCard route={route} dispatch={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("returns null if route has no schedules", () => {
  const routeWithoutSchedules: RouteWithStopsWithDirections = {
    ...route,
    stops_with_directions: [
      {
        stop,
        directions: [],
        distance: "1 mi"
      }
    ]
  };
  createReactRoot();
  const tree = renderer
    .create(<RouteCard route={routeWithoutSchedules} dispatch={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("it renders a stop card for the silver line", () => {
  createReactRoot();
  const sl: RouteWithStopsWithDirections = {
    route: { ...route.route, id: "751" },
    stops_with_directions: route.stops_with_directions
  };
  const tree = renderer
    .create(<RouteCard route={sl} dispatch={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
