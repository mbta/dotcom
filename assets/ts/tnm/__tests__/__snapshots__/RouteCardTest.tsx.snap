// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a stop card 1`] = `
<div
  className="m-tnm-sidebar__route"
  data-mode="bus"
>
  <div
    className="c-link-block h3 m-tnm-sidebar__route-name u-bg--bus"
  >
    <a
      className="c-link-block__outer-link"
      href="/schedules/route-id"
    >
      <span
        className="sr-only"
      >
        Go to route
      </span>
    </a>
    <div
      className="c-link-block__inner"
    >
      <span
        className="notranslate"
      >
        <span
          className="bus-route-sign"
        >
          Route Name
        </span>
        <div
          className="c-link-block__view-schedule"
        >
          View schedule
        </div>
      </span>
    </div>
  </div>
  <div
    className="m-tnm-sidebar__route-stop"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    style={
      Object {
        "cursor": "pointer",
      }
    }
    tabIndex={0}
  >
    <div
      className="m-tnm-sidebar__stop-info"
    >
      <a
        className="m-tnm-sidebar__stop-name notranslate"
        href="/stops/stop-id"
      >
        <span
          aria-hidden="true"
          className="notranslate m-tnm-sidebar__stop-marker"
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
        Stop Name
        <span
          aria-hidden="false"
          className="notranslate m-tnm-sidebar__stop-accessible"
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </a>
      <div
        className="m-tnm-sidebar__stop-distance"
      >
        238 ft
      </div>
    </div>
    <div
      className="m-tnm-sidebar__direction"
    >
      <div
        className="m-tnm-sidebar__direction-name u-small-caps"
      >
        Outbound
      </div>
      <div
        className="m-tnm-sidebar__direction-headsigns"
      >
        <div
          className="m-tnm-sidebar__headsign-schedule m-tnm-sidebar__headsign-schedule--condensed"
        >
          <div
            className="m-tnm-sidebar__headsign"
          >
            <div
              className="m-tnm-sidebar__headsign-name m-tnm-sidebar__headsign-name--large notranslate"
            >
              Headsign
            </div>
          </div>
          <div
            className="m-tnm-sidebar__schedules"
          >
            <div
              className="m-tnm-sidebar__schedule"
            >
              <div
                className="m-tnm-sidebar__time m-tnm-sidebar__time--schedule"
              >
                <div
                  className="m-tnm-sidebar__time-number"
                >
                  4:30
                </div>
                <div
                  className="m-tnm-sidebar__time-mins"
                >
                  PM
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`it renders a stop card for the silver line 1`] = `
<div
  className="m-tnm-sidebar__route"
  data-mode="bus"
>
  <div
    className="c-link-block h3 m-tnm-sidebar__route-name u-bg--silver-line"
  >
    <a
      className="c-link-block__outer-link"
      href="/schedules/751"
    >
      <span
        className="sr-only"
      >
        Go to route
      </span>
    </a>
    <div
      className="c-link-block__inner"
    >
      <span
        className="notranslate"
      >
        <span
          className=""
        >
          Silver Line Route Name
        </span>
        <div
          className="c-link-block__view-schedule"
        >
          View schedule
        </div>
      </span>
    </div>
  </div>
  <div
    className="m-tnm-sidebar__route-stop"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    style={
      Object {
        "cursor": "pointer",
      }
    }
    tabIndex={0}
  >
    <div
      className="m-tnm-sidebar__stop-info"
    >
      <a
        className="m-tnm-sidebar__stop-name notranslate"
        href="/stops/stop-id"
      >
        <span
          aria-hidden="true"
          className="notranslate m-tnm-sidebar__stop-marker"
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
        Stop Name
        <span
          aria-hidden="false"
          className="notranslate m-tnm-sidebar__stop-accessible"
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </a>
      <div
        className="m-tnm-sidebar__stop-distance"
      >
        238 ft
      </div>
    </div>
    <div
      className="m-tnm-sidebar__direction"
    >
      <div
        className="m-tnm-sidebar__direction-name u-small-caps"
      >
        Outbound
      </div>
      <div
        className="m-tnm-sidebar__direction-headsigns"
      >
        <div
          className="m-tnm-sidebar__headsign-schedule m-tnm-sidebar__headsign-schedule--condensed"
        >
          <div
            className="m-tnm-sidebar__headsign"
          >
            <div
              className="m-tnm-sidebar__headsign-name m-tnm-sidebar__headsign-name--large notranslate"
            >
              Headsign
            </div>
          </div>
          <div
            className="m-tnm-sidebar__schedules"
          >
            <div
              className="m-tnm-sidebar__schedule"
            >
              <div
                className="m-tnm-sidebar__time m-tnm-sidebar__time--schedule"
              >
                <div
                  className="m-tnm-sidebar__time-number"
                >
                  4:30
                </div>
                <div
                  className="m-tnm-sidebar__time-mins"
                >
                  PM
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`returns null if route has no schedules 1`] = `null`;
